/******************************************************************************
 * @file    LteDirectDiscovery.proto
 * @brief   This proto file describes the interface between RIL and
 *          LteDirectDiscovery service
 *
 * ---------------------------------------------------------------------------
 * Copyright (c) 2016 Qualcomm Technologies, Inc.
 * All Rights Reserved.
 * Confidential and Proprietary - Qualcomm Technologies, Inc.
 * ---------------------------------------------------------------------------
 *
 *******************************************************************************/

/* LteDirectDiscovery interface to RIL */
package LteDirectDiscovery;
option java_package = "com.qualcomm.qti.ltedirect";
option java_outer_classname = "LteDirectDiscoveryIF";


// RIL Discovery message
message MsgHeader {
    required fixed32 token = 1; /* generated dynamically */
    required MsgType type = 2;
    required MsgId id = 3;
    required Error error = 4;
    optional bytes payload = 5;
}

enum MsgType {
    UNKNOWN = 0;
    REQUEST = 1;
    RESPONSE = 2;
    UNSOL_RESPONSE = 3;
}

enum Error{
    E_SUCCESS = 0;
    E_RADIO_NOT_AVAILABLE = 1;     /* If radio did not start or is resetting */
    E_GENERIC_FAILURE = 2;
    E_NOT_SUPPORTED = 3;
    E_APP_AUTH_FAILURE = 4;
}

enum MsgId {
    REQUEST_BASE = 0;

    /**
     * MsgType: REQUEST,
     * MsgId:   REQUEST_INITIALIZE,
     * Error:   E_UNUSED,
     * Message: NULL
     *
     * MsgType: RESPONSE,
     * MsgId:   REQUEST_INITIALIZE,
     * Error:   Valid errors,
     * Message: NULL
     *
     * Description: Cancel the subscription
     *
     * Valid errors:
     *  E_SUCCESS
     *  E_RADIO_NOT_AVAILABLE
     *  GENERIC_FAILURE
     */
    REQUEST_INITIALIZE = 1;

    /**
     * MsgType: REQUEST,
     * MsgId:   REQUEST_GET_DEVICE_CAPABILITY,
     * Error:   E_UNUSED,
     * Message: NULL
     *
     * MsgType: RESPONSE,
     * MsgId:   REQUEST_GET_DEVICE_CAPABILITY,
     * Error:   Valid errors,
     * Message: message DeviceCapability
     *
     * Description: Get device LTE-D capability of each subscriptions
     *
     * Valid errors:
     *  E_SUCCESS
     *  E_RADIO_NOT_AVAILABLE
     *  GENERIC_FAILURE
     */
    REQUEST_GET_DEVICE_CAPABILITY = 2;

    /**
     * MsgType: REQUEST,
     * MsgId:   REQUEST_TERMINATE,
     * Error:   E_UNUSED,
     * Message: message Terminate
     *
     * MsgType: RESPONSE,
     * MsgId:   REQUEST_TERMINATE,
     * Error:   Valid errors,
     * Message: NULL
     *
     * Description: Stop all ongoing operations
     *
     * Valid errors:
     *  E_SUCCESS
     *  E_RADIO_NOT_AVAILABLE
     *  GENERIC_FAILURE
     */
    REQUEST_TERMINATE = 3;

    /**
     * MsgType: REQUEST,
     * MsgId:   REQUEST_GET_SERVICE_STATUS,
     * Error:   E_UNUSED,
     * Message: NULL
     *
     * MsgType: RESPONSE,
     * MsgId:   REQUEST_GET_SERVICE_STATUS,
     * Error:   Valid errors,
     * Message: message ServiceStatus
     *
     * Description: Get publish and subscribe service status
     *
     * Valid errors:
     *  E_SUCCESS
     *  E_RADIO_NOT_AVAILABLE
     *  GENERIC_FAILURE
     */
    REQUEST_GET_SERVICE_STATUS = 4;

    /**
     * MsgType: REQUEST,
     * MsgId:   REQUEST_PUBLISH,
     * Error:   E_UNUSED,
     * Message: message Publish
     *
     * MsgType: RESPONSE,
     * MsgId:   REQUEST_PUBLISH,
     * Error:   Valid errors,
     * Message: NULL
     *
     * Description: Publish an expression
     *
     * Valid errors:
     *  E_SUCCESS
     *  E_RADIO_NOT_AVAILABLE
     *  GENERIC_FAILURE
     */
    REQUEST_PUBLISH = 5;

    /**
     * MsgType: REQUEST,
     * MsgId:   REQUEST_CANCEL_PUBLISH,
     * Error:   E_UNUSED,
     * Message: message CancelPublish
     *
     * MsgType: RESPONSE,
     * MsgId:   REQUEST_CANCEL_PUBLISH,
     * Error:   Valid errors,
     * Message: NULL
     *
     * Description: Cancel the publication of an expression
     *
     * Valid errors:
     *  E_SUCCESS
     *  E_RADIO_NOT_AVAILABLE
     *  GENERIC_FAILURE
     */
    REQUEST_CANCEL_PUBLISH = 6;

    /**
     * MsgType: REQUEST,
     * MsgId:   REQUEST_SUBSCRIBE,
     * Error:   E_UNUSED,
     * Message: message Subscribe
     *
     * MsgType: RESPONSE,
     * MsgId:   REQUEST_SUBSCRIBE,
     * Error:   Valid errors,
     * Message: NULL
     *
     * Description: Subscribe an expression
     *
     * Valid errors:
     *  E_SUCCESS
     *  E_RADIO_NOT_AVAILABLE
     *  GENERIC_FAILURE
     */
    REQUEST_SUBSCRIBE = 7;

    /**
     * MsgType: REQUEST,
     * MsgId:   REQUEST_CANCEL_SUBSCRIBE,
     * Error:   E_UNUSED,
     * Message: message CancelSubscribe
     *
     * MsgType: RESPONSE,
     * MsgId:   REQUEST_CANCEL_SUBSCRIBE,
     * Error:   Valid errors,
     * Message: NULL
     *
     * Description: Cancel the subscription
     *
     * Valid errors:
     *  E_SUCCESS
     *  E_RADIO_NOT_AVAILABLE
     *  GENERIC_FAILURE
     */
    REQUEST_CANCEL_SUBSCRIBE = 8;

    UNSOL_RSP_BASE = 1000;

    /**
     * MsgType: UNSOL_RESPONSE,
     * MsgId:   UNSOL_RESPONSE_DEVICE_CAPABILITY_CHANGED,
     * Error:   E_UNUSED,
     * Message: message DeviceCapability
     *
     * Description: Notifies the device LTE-D capability of all SUBs.
     */
    UNSOL_RESPONSE_DEVICE_CAPABILITY_CHANGED = 1001;

    /**
     * MsgType: UNSOL_RESPONSE,
     * MsgId:   UNSOL_RESPONSE_SERVICE_STATUS,
     * Error:   E_UNUSED,
     * Message: message ServiceStatus
     *
     * Description: Notifies the LTE-D service status
     */
    UNSOL_RESPONSE_SERVICE_STATUS = 1002;

    /**
     * MsgType: UNSOL_RESPONSE,
     * MsgId:   UNSOL_RESPONSE_AUTHORIZATION_RESULT,
     * Error:   E_UNUSED,
     * Message: message AuthorizationResult
     *
     * Description: Notifies App authorization failure from LTE-D server
     */
    UNSOL_RESPONSE_AUTHORIZATION_RESULT = 1003;

    /**
     * MsgType: UNSOL_RESPONSE,
     * MsgId:   UNSOL_RESPONSE_EXPRESSION_STATUS,
     * Error:   E_UNUSED,
     * Message: message ExpressionStatus
     *
     * Description: Notifies the status of the Expressions
     */
    UNSOL_RESPONSE_EXPRESSION_STATUS = 1004;

    /**
     * MsgType: UNSOL_RESPONSE,
     * MsgId:   UNSOL_RESPONSE_TRANSMISSION_STATUS,
     * Error:   E_UNUSED,
     * Message: message TransmissionStatus
     *
     * Description: Transmission Status
     */
    UNSOL_RESPONSE_TRANSMISSION_STATUS = 1005;

    /**
     * MsgType: UNSOL_RESPONSE,
     * MsgId:   UNSOL_RESPONSE_RECEPTION_STATUS,
     * Error:   E_UNUSED,
     * Message: message ReceptionStatus
     *
     * Description: Reception Status
     */
    UNSOL_RESPONSE_RECEPTION_STATUS = 1006;

    /**
     * MsgType: UNSOL_RESPONSE,
     * MsgId:   UNSOL_RESPONSE_MATCH_EVENT,
     * Error:   E_UNUSED,
     * Message: message MatchEvent
     *
     * Description: Notifies the Match Event
     */
    UNSOL_RESPONSE_MATCH_EVENT = 1007;
}

message DeviceCapability {
    optional fixed32 capability = 2; /* FALSE (0) - Disabled */
                                     /* TRUE (1)  - Enabled */
}

message Terminate {
    optional string osAppId = 1;
}

message ServiceStatus {
    optional fixed32 publishAllowed = 2;
    optional fixed32 subscribeAllowed = 3;
}

message Publish {
    optional string osAppId = 1;
    optional string expression = 2;
    optional fixed32 expressionValidityTime = 3;
    optional string metaData = 4;
    optional DiscoveryType discoveryType = 5;
    optional fixed32 duration = 6;
}

message CancelPublish {
    optional string osAppId = 1;
    optional string expression = 2;
}

message Subscribe {
    optional string osAppId = 1;
    optional string expression = 2;
    optional fixed32 expressionValidityTime = 3;
    optional DiscoveryType discoveryType = 4;
    optional fixed32 duration = 5;
}

message CancelSubscribe {
    optional string osAppId = 1;
    optional string expression = 2;
}

message AuthorizationResult {
    optional string osAppId = 1;
    optional Result result = 2;
}

message ExpressionStatus {
    optional string osAppId = 1;
    optional string expression = 2;
    optional Result result = 3;
}

message TransmissionStatus {
    optional string osAppId = 1;
    optional string expression = 2;
    optional fixed32 status = 3;
}

message ReceptionStatus {
    optional string osAppId = 1;
    optional string expression = 2;
    optional fixed32 status = 3;
}

message MatchEvent {
    optional string osAppId = 1;
    optional string expression = 2;
    optional string matchedExpression = 3;
    optional fixed32 state = 4;
    optional fixed32 metaDataIndex = 5;
    optional string metaData = 6;
}

enum DiscoveryType {
    INVALID = 0;
    OPEN = 1;
    RESTRICTED = 2;
}

enum Result {
    SUCCESS = 0;
    GENERIC_FAILURE = 1;
    IN_PROGRESS = 2;
    INVALID_EXPRESSION_SCOPE = 3;
    UNKNOWN_EXPRESSION = 4;
    INVALID_DISCOVERY_TYPE = 5;
    SERVICE_NOT_AVAILABLE = 6;
    APP_AUTH_FAILURE = 7;
    NOT_SUPPORTED = 8;
}
